src/com/github/Kraken3/AFKPGC/BotDetector.java

```diff
@@ -0,0 +1,306 @@
+								truebot, "): ", bounds);
+					} else {
+						// else not enough info yet. Pass.
+						AFKPGC.debug("Player ", curSuspect.getUUID(), " Not enough bounds data, skip for now");
+					}
+				}
+			} else {
+				AFKPGC.debug("No suspects this round.");
+			}
+
+			if (lastRoundSuspect != null) {
+				// TODO: This is starting to stink of synchronization issues. We shouldn't need
+				//       to check immune so often within the same compute round.
+				if (!AFKPGC.immuneAccounts.contains(lastRoundSuspect.getUUID())) {
+					if (currentTPS - lastRoundTPS > criticalTPSChange) {
+						/* This can be relatively sensitive, because it will only ban players for 
+						 * a longer period of time, if it catches them twice here */
+						Date currentDate = new Date();
+						if (suspectedBotters.contains(lastRoundSuspect.getName())) {
+							if (longBans) {
+								BanEntry leBan = banList.addBan( lastRoundSuspect.getName(),
+										"Kicking you resulted in a noticeable TPS improvement, so you " +
+										"were banned until the TPS goes back to normal values.",
+										new Date(currentDate.getTime() + longBan),
+										null); // long ban.
+								Player p = Bukkit.getPlayer(lastRoundSuspect.getUUID());
+								if (p != null) {
+					   				p.kickPlayer(leBan.getReason());
+								}
+								bannedPlayers.add(lastRoundSuspect.getName());
+								addToBanfile(lastRoundSuspect.getName());
+								suspectedBotters.remove(lastRoundSuspect.getName());
+								AFKPGC.debug("Player ", lastRoundSuspect.getUUID(), " long banned for ",
+										longBan," confirmed lag source.");
+							}
+							AFKPGC.logger.info("The player " + lastRoundSuspect.getName()
+									+ " causes lag and is a repeated offender, kicking him resulted"
+									+ " in a TPS improvement of " + String.valueOf(currentTPS - lastRoundTPS)
+									+ " at the location " + lastRoundSuspect.getLocation().toString());
+
+						} else {
+							suspectedBotters.add(lastRoundSuspect.getName());
+							AFKPGC.logger.info( "The player " + lastRoundSuspect.getName()
+									+ " is suspected to cause lag, kicking him resulted in a TPS improvement of "
+									+ String.valueOf(currentTPS - lastRoundTPS) + " at the location "
+									+ lastRoundSuspect.getLocation().toString());
+						}
+					} else {
+						AFKPGC.debug("Player ", lastRoundSuspect.getUUID(), " (", lastRoundSuspect.getName(),
+								") cleared, kicking them did not improve TPS -- granting a reprieve");
+						reprieve.put(lastRoundSuspect.getUUID(), maxReprieve);
+					}
+				} else {
+					AFKPGC.debug("Player ", lastRoundSuspect.getUUID(), " was suspected but is immune.");
+				}
+			} else {
+				AFKPGC.debug("No prior suspects to evaluable.");
+			}
+
+			lastRoundTPS = currentTPS;
+			lastRoundSuspect = thisRoundSuspect;
+		} else { // TPS is high enough
+			// TODO: Evaluate if this makes sense. We should probably see if TPS stays good for a while

```

Comment:
We could just keep the players causing lag locked away for 30 minutes after the tick is good again, that sounds reasonable imo.